pkg/acsclient/utils.go

```diff
@@ -0,0 +1,98 @@
+package acsclient
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"golang.org/x/net/http2"
+)
+
+var (
+	defaultTimeOut = 50
+	defaultRetry   = 5
+)
+
+type requestParams struct {
+	method           string
+	url              string
+	timeout          int
+	token            string
+	isStreamResponse bool
+	resultKeyword    string
+	retry            int
+}
+
+func getStreamResponse(keyword string, resp *http.Response, responseObject interface{}) error {
+	responseData := map[string][]interface{}{
+		"results": make([]interface{}, 0),
+	}
+	jdecoder := json.NewDecoder(resp.Body)
+	for jdecoder.More() {
+		data := make(map[string]map[string]interface{})
+		err := jdecoder.Decode(&data)
+		if err != nil {
+			err = fmt.Errorf("failed to decode :%v", err)
+			return err
+		}
+		responseData["results"] = append(responseData["results"], data["result"][keyword])
+	}
+	bytesData, err := json.Marshal(responseData)
+	if err != nil {
+		return fmt.Errorf("failed to marshal data : %v", err)
+	}
+	return json.Unmarshal(bytesData, &responseObject)
+}
+func makeRequest(requestParams requestParams, responseObject interface{}) error {
+	timeout := defaultTimeOut
+	if requestParams.timeout != 0 {
+		timeout = requestParams.timeout
+	}
+	retries := defaultRetry
+	if requestParams.retry != 0 {
+		retries = requestParams.retry
+	}
+	transport := &http2.Transport{
+		TLSClientConfig: &tls.Config{
+			MinVersion: tls.VersionTLS13,
+		},
+		DialTLS: func(network, addr string, cfg *tls.Config) (net.Conn, error) {
+			return tls.DialWithDialer(&net.Dialer{Timeout: time.Duration(requestParams.timeout) * time.Second}, network, addr, cfg)
+		},
+	}
+	client := retryablehttp.NewClient()
```

Comment:
will this create a new http client on every requests, is there a way to reuse connection pool?